File: communication/src/main/java/net/tomofiles/skysign/communication/domain/mission/Mission.java
package net.tomofiles.skysign.communication.domain.mission;

import java.util.ArrayList;
import java.util.List;

import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@EqualsAndHashCode(of = {"id"})
@ToString
public class Mission {
    @Getter
    private final MissionId id;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private List<Waypoint> waypoints;

    public Mission(MissionId id) {
        this.id = id;
        this.waypoints = new ArrayList<>();
    }

    public int pushWaypoint(
        double latitudeDegree, 
        double longitudeDegree, 
        double relativeHeightM, 
        double speedMS) {
            int order = this.waypoints.size() + 1;
            this.waypoints.add(
                new Waypoint(
                    order, 
                latitudeDegree, 
                longitudeDegree,
                 relativeHeightM, 
                 speedMS)
            );
            return order;
    }
}

File: remote-communication/pkg/mission/domain/mission/mission.go
package mission

// ID .
type ID string

// Waypoint .
type Waypoint struct {
	PointOrder        int
	LatitudeDegree    float64
	LongitudeDegree   float64
	RelativeAltitudeM float64
	SpeedMS           float64
}

// Mission .
type Mission struct {
	id        ID
	waypoints []*Waypoint
}

// GetID .
func (m *Mission) GetID() ID {
	return m.id
}

// PushWaypoint .
func (m *Mission) PushWaypoint(
	latitudeDegree float64,
	longitudeDegree float64,
	relativeAltitudeM float64,
	speedMS float64,
) int {
	order := len(m.waypoints) + 1
	m.waypoints = append(
		m.waypoints,
		&Waypoint{
			PointOrder:        order,
			LatitudeDegree:    latitudeDegree,
			LongitudeDegree:   longitudeDegree,
			RelativeAltitudeM: relativeAltitudeM,
			SpeedMS:           speedMS,
		},
	)
	return order
}

// GetWaypoints .
func (m *Mission) GetWaypoints() []*Waypoint {
	return m.waypoints
}
